File: packages/sdk/typescript/human-protocol-sdk/src/graphql/queries/payout.ts
import gql from 'graphql-tag';
import { IPayoutFilter } from 'src/interfaces';

const PAYOUT_FRAGMENT = gql`
  fragment PayoutFields on Payout {
    id
    escrowAddress
    recipient
    amount
    createdAt
  }
`;

export const GET_PAYOUTS_QUERY = (filter: IPayoutFilter) => {
  const { escrowAddress, recipient, from, to } = filter;

  const WHERE_CLAUSE = `
    where: {
      ${escrowAddress ? `escrowAddress: $escrowAddress` : ''}
      ${recipient ? `recipient: $recipient` : ''}
      ${from ? `createdAt_gte: $from` : ''}
      ${to ? `createdAt_lt: $to` : ''}
    }
  `;

  return gql`
    query getPayouts(
      $escrowAddress: String
      $recipient: String
      $from: Int
      $to: Int
      $first: Int
      $skip: Int
      $orderDirection: String
    ) {
      payouts(
        ${WHERE_CLAUSE}
        orderBy: createdAt,
        orderDirection: $orderDirection,
        first: $first,
        skip: $skip
      ) {
        ...PayoutFields
      }
    }
    ${PAYOUT_FRAGMENT}
  `;
};

File: packages/sdk/python/human-protocol-sdk/human_protocol_sdk/gql/payout.py
from human_protocol_sdk.filter import PayoutFilter

payout_fragment = """
fragment PayoutFields on Payout {
    id
    escrowAddress
    recipient
    amount
    createdAt
}
"""


def get_payouts_query(filter: PayoutFilter) -> str:
    return """
query GetPayouts(
    $escrowAddress: String,
    $recipient: String,
    $from: Int,
    $to: Int,
    $first: Int,
    $skip: Int,
    $orderDirection: String
) {{
    payouts(
      where: {{
        {escrow_address_clause}
        {recipient_clause}
        {from_clause}
        {to_clause}
      }}
      orderBy: createdAt
      orderDirection: $orderDirection
      first: $first
      skip: $skip
    ) {{
      ...PayoutFields
    }}
}}
{payout_fragment}
""".format(
        payout_fragment=payout_fragment,
        escrow_address_clause=(
            "escrowAddress: $escrowAddress" if filter.escrow_address else ""
        ),
        recipient_clause="recipient: $recipient" if filter.recipient else "",
        from_clause="createdAt_gte: $from" if filter.date_from else "",
        to_clause="createdAt_lte: $to" if filter.date_from else "",
    )
